### DISCUSSION TITLE
Authenticated App - how do I associate records with user?

### DISCUSSION THREAD

**Discussion by knaopel** [Azure/azure-mobile-apps#590] [Q&A] [answered]
I have some old Azure Mobile Services app that I want to upgrade. I used the following code in my Node.JS backend to record the user with my table rows:
```
// CREATE operation
table.insert(function (context) {
  context.item.userId = context.user.id;
  return context.execute();
});
 
// READ operation
table.read(function (context) {
  context.query.where({ userId: context.user.id });
  return context.execute();
});
```
Now I am using the ASP.NET 6 backend with the Microsoft.AspNetCore.DataSync and Microsoft.AspNetCore.DataSync.EFCore packages. I have figured out how to override the CreateAsync method like below:
```
public override Task<IActionResult> CreateAsync([FromBody] TimecardEntryItem item, CancellationToken token = default)
{
  return base.CreateAsync(item, token);
}
```
The question is how do I inject the user Id from the authentication into the CreateAsync method?

**Comment by adrianhall** [comment, marked as answer]
https://learn.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#configure-access-permissions

**Reply by adrianhall** [reply]
It's literally the example provided.

**Reply by knaopel** [reply]
Pardon my ignorance, @adrianhall but what is the IUserId in this example? I am getting a compiler error on the signature for the class:
```
public class PrivateAccessControlProvider<T> : IAccessControlProvider<T>
    where T : ITableData
    where T : IUserId
```
cs0409 T cannot have multiple constraint clauses. https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0409

**Reply by adrianhall** [reply]
IUsedId is simply an interface with `string UserId { get; }` in it. You can adjust by the following:

``` csharp
where T : class, ITableData, IUserId
```

I'll get the docs updated with the proper syntax today.

**Reply by knaopel** [reply]
Hey @adrianhall _accessor.HttpContext.User?.Identity?.Name in the getter of UserId is always null, even when I am authenticated. Any clues on how to get that populated or where else I can get the user?

**Reply by adrianhall** [reply]
The identity provider will fill that in.  You should take a look at Identity and User of the HttpContext when you are authenticated and see what is available.

**Reply by knaopel** [reply]
As you can see there is no information in User.Identity even though "IsAuthenticated" is true.
![image](https://github.com/Azure/azure-mobile-apps/assets/1695947/f43dd40b-97b6-44da-bff5-94b194e8b363)

**Reply by adrianhall** [reply]
There are 18 things in Claims.  I suspect that the NameClaimType is not among them, which is why Name is not set.

This suggests that your ASP.NET Core Authentication and Authorization setup is wrong.  How it is wrong, I don't know.  I'm not an expert.  This, however, has nothing to do with Azure Mobile Apps - we encourage the use of the standard stuff rather than inventing our own.

**Reply by knaopel** [reply]
Fair enough. Thanks for the help.
